keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+- Kubernetes 1.13 - The Application Kind is promoted to the apps/v1 Group Version and deprecated in the apps/v1beta1 
+Group Version.
+- Kubernetes 1.14 - Storage migration is complete, on upgrade, to the apps/v1 Group Version of the Application Kind and 
+this version is disabled by default.
+- Kuberentes 1.15 - The apps/v1beta1 version of the Application Kind is removed from the API.
+
+## Drawbacks
+
+### Increased Memory, CPU, and Network Requirements
+The Application Controller, as designed, like the Garbage Collector, must watch a large fraction of the API objects 
+stored on the API server. We must be careful to implement the Application Controller efficiently to minimize 
+the increased memory and CPU requirements of the core Controller Manager and to minimize pressure on the API server. 
+
+### Increased API Server Storage Requirements
+As the Application object will be stored on the API Server, if the users create many Applications, the API Server 
+(i.e. etcd) will have an increased storage requirement. Users already create many ConfigMaps and Secrets. It is worth
+pointing out that if Application metadata is stored on every object using label and annotation conventions, the 
+storage burden on the API Server will likely be greater than if we aggregate the information in a single object and 
+store it once. Also, users are already using CRDs for a similar function, and, for those users, the API Server already 
+tolerates the additional storage burden.
+
+## Alternatives 
+This section contains various design alternatives that were considered during the generation of this KEP.
+
+### Use Labeling Conventions
+Rather than introducing an API object, we could require that all tools use a standard labeling scheme to identify 
+the components of an Application. This would also provide an opt-in method for tool interoperability. However, this 
+method provides no way to garbage collect Application components, express dependencies between applications, or to 
+aggregate application meta-data in a single, well-known object. Also, there is no way to ensure that tools implement 
+labeling conventions consistently or correctly. Well intentioned implementors may accidentally produce tools that 
+incorrectly implement the convention leading to degraded or broken user experiences for other tools or UIs.
+This method lacks the capacity for server side validation of the applications definition. There is not way to 
+ensure a collection of labels correctly communicates the users intentions. Lastly, this method lacks any 
+method to provide health checks.
+
+### Use Custom Resource Definitions
+Rather than introducing an API object, each tool could simply use a CRD (Custom Resource Definition) to represent 
+Applications. As CRDs are, by definition, a custom API extension mechanism, it is unlikely that this method would 
+provide for the desired level of interoperability across tools and UIs. If each tool uses their own CRD, they will a 
```

Comment:
I'm of the opinion that this should start as a CRD, for similar reasons.

Reply:
I'm with @smarterclayton on this: Why not just implement _this spec_ as a CRD, and then push for widespread adoption? There's nothing about CRDs as a technology that causes us to have multiple implementations. It's merely that no single abstraction has yet surfaced as a reasonable solution to the problem.

The argument that "If all tools use a common CRD..." also doesn't follow. It is 100% reasonable to have an oft-used or even ubiquitous Kind that is not in core. In fact, if Kubernetes is to succeed as a platform, we _have_ to get to that point. Otherwise we'll succumb to the same bloat-induced fate of other projects.

On a related note, given that I believe the controller will be more complicated than the proposal seems to indicate, I would hate for "interoperability with Application" to be a new gate for other candidate core features. If it is kept as a separate project, then it can be grown independently of the main Kubernetes source, which is virtuous in this case.

I personally would rather see this implemented as a stand-alone entity that could perhaps be incubated into core if at some point it makes sense. But at this point, I feel like putting it into the core would be a disingenuous attempt to "kill off the competition" before really showing this idea actually meets the needs of Kubernetes operators. Please don't misread this as me saying I don't like the proposal or the idea -- just that I don't see merit in pushing this straight to core.

Reply:
I should change the wording. It appears I did not clearly communicate my point.
As several organizations are concurrently working on (or have deployed in production) several CRDs that attempt to solve the same problem, we're proposing that it is appropriate to aggregate the lessons learned and design a solution in tree that meets the needs of the community in general.

> I would hate for "interoperability with Application" to be a new gate for other candidate core features. 

This must, by design, never be the case. As stated in the goals section it must be a strictly opt-in solution.

> I personally would rather see this implemented as a stand-alone entity that could perhaps be incubated into core if at some point it makes sense. But at this point, I feel like putting it into the core would be a disingenuous attempt to "kill off the competition"

That should be listed as a goal ... actually perhaps we should have a section called "anti-goals". The reason we are proposing an alpha resource in apps is to leverage the community and infrastructure (e.g. testing, code generation, server side validation, resource versioning) that exists in Kubernetes core.

Based on initial feedback, imo there are enough contributors and interest that we could produce something reasonable, at an alpha level of stability, in the mian repo. If the community as whole feels that this is not the case, then we could consider creating a sub-project and continuing to collaborate across organizations to develop a CRD and third party controller there. Aside from the issues pointed out [below](https://github.com/kubernetes/community/pull/1629/), we would largely be operating outside the existing governance of the Kubernetes project. If we think that we should never add this proposal to core, then perhaps that is the right direction.

However,  if we think we eventually want a GA Application object in the apps group, if we want to design and implement it as a community, and if we want to ensure that Kubernetes governance is applied to establishing the maturity of the resource as it is promoted through various levels of stability, the process that has precedence is starting with an alpha resource.

To be clear, I'm not saying I disagree with starting as a CRD, but  we should carefully consider what we trade if we take this approach. In the same way that OpenShift still supports DeploymentConfig and Deployment, should we eventually add an apps Application resource, we would likely have to support both the CRD and the multiple versions of the core resource for quite some time (here I'm thinking in terms of those organizations that support many Kubernetes clusters at different versions simultaneously). 

> I don't follow this line of reasoning. Regular APIs aren't special. If a CRD can't accomplish interop, we will iterate on CRD. There are downsides to CRD, but huge upsides, like you can iterate on them rapidly and demonstrate value before dropping a massive new thing in Kubernetes core.

Agree. To me the benefit of in-tree development (as above) is established governance, existing tooling, and an existing repo where we can iterate (potentially more slowly). Shipping the right thing is more important than shipping something (imo).